pkg/cloud/services/loadbalancer/loadbalancer.go

```diff
@@ -658,32 +665,86 @@ func (s *Service) ReconcileLoadBalancerMember(openStackCluster *infrav1.OpenStac
 	return nil
 }
 
-func (s *Service) DeleteLoadBalancer(openStackCluster *infrav1.OpenStackCluster, clusterResourceName string) error {
+func (s *Service) checkIfLbPortsExist(openStackCluster *infrav1.OpenStackCluster) (bool, error) {
+	// If the LB status doesn't exist, we can't get the LB ID so it's safer to just
+	// assume the port can't be found anymore.
+	// We could get the LB name and gets its ID but if the LB was deleted, we would
+	// get an error and we would have to handle it.
+	if openStackCluster.Status.APIServerLoadBalancer == nil {
+		return false, nil
+	}
+
+	// If the network is not ready, do nothing
+	if openStackCluster.Status.Network == nil || openStackCluster.Status.Network.ID == "" {
+		return false, nil
+	}
+	networkID := openStackCluster.Status.Network.ID
+	lbID := openStackCluster.Status.APIServerLoadBalancer.ID
+
+	// When creating a LoadBalancer, Octavia will create two ports but only
+	// one of them has a device ID which is the LoadBalancer ID.
+	// We can't easily identify the other one beside its name starting with "octavia-lb-".
+	// So for now we don't handle this port as another cluster could live in this network
+	// which would lead to a conflict.
+	portOpts := ports.ListOpts{
+		NetworkID: networkID,
+		DeviceID:  "lb-" + lbID,
+	}
+
+	portList, err := s.networkingService.GetPortWithPrefix(portOpts, loadBalancerPortPrefix)
+	if err != nil {
+		return false, err
+	}
+	if len(portList) > 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (s *Service) DeleteLoadBalancer(openStackCluster *infrav1.OpenStackCluster, clusterResourceName string) (result *ctrl.Result, reterr error) {
 	loadBalancerName := getLoadBalancerName(clusterResourceName)
 	lb, err := s.checkIfLbExists(loadBalancerName)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	lbPortsExist, err := s.checkIfLbPortsExist(openStackCluster)
+	if err != nil {
+		return nil, err
 	}
 
 	if lb == nil {
-		return nil
+		if lbPortsExist {
+			s.scope.Logger().Info("Load balancer ports still exist, waiting for them to be deleted", "name", loadBalancerName)
```

Comment:
Looking at the [logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/2122/pull-cluster-api-provider-openstack-e2e-test/1800982517210157056/artifacts/clusters/bootstrap/logs/capo-system/capo-controller-manager/capo-controller-manager-7d957cb6cd-dhmz7/manager.log), it seems like we never hit that so I wonder if the [requeue](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2122/files#diff-29cb3ac9349ca04685d8b48d0b1bc13a90e76242d0300fea7ed14f4f71ef2eecR769) at the end of the function has given enough time to perform the port cleanup on the Octavia side.

Reply:
See below. This isn't a case of 'enough time'; we're now waiting until the LB is gone, however long that takes. We don't need this check.